### PR TITLE
Added pre-commit hooks

### PR DESCRIPTION
**Pull request by Zabuzard**
## Overview

This adds a pre-commit hook with an automatic installation to get rid of

> Fuck you Spotless

commits 😆

Gradle installs the hook and the hook applies spotless ~~and runs the unit tests~~.

Downside, commits now take about ~~10 seconds~~ 2 seconds.

### CODE REVIEW

**Review by Tais993 — DISMISSED** [review]
I rather wait 10s than manually run spotless & commit & squash & force-push & hate Spotless :p

**Review by Tais993 — DISMISSED** [review]

**Comment by borgrel** [comment]
cant u get it to apply spotless with github actions?

**Comment by Zabuzard** [comment]
@borgrel That would be pretty messy. Since then someone else creates new commits on "your branch", which interfers with your local branch. I.e. you have to do `git pull` each time after a `git push` then. And we pollute our git history then with "Spotless apply" extra commits after each commit. And if you let it modify the existing commit, then its even worse for the user, since their local branch is then incompatible with the remote branch.

Everyone and their mother uses pre-hooks for anything that actually changes code/commits for that reason.

**Comment by borgrel** [comment]
nods

thanks for the explination

File: scripts/pre-commit [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+if ! ./gradlew spotlessApply; then
+  # Restore unstaged changes
+  git stash pop -q
+  exit $?
+fi
+# Add Spotless changes
+git add -A
+
+# Restore unstaged changes
+git stash pop -q
```
[the remark this answers is not in the document]
Comment (marko-radosavljevic) [reply scripts/pre-commit:19]: Yeah

> As noted in a comment below, the final git stash pop --index -q complains a bit (or, worse, restores an old stash) if the initial git stash push step finds no changes to save. You should therefore protect the "restore" step with a test to see if the "save" step actually stashed anything.
>
> The initial git stash --keep-index -q simply exits quietly (with status 0) when it does nothing, so we need to handle two cases: no stash exists either before or after the save; and, some stash existed before the save, and the save did nothing so the old existing stash is still the top of the stash stack.

https://stackoverflow.com/questions/20479794/how-do-i-properly-git-stash-pop-in-pre-commit-hooks-to-get-a-clean-working-tree

https://www.darrenlester.com/blog/git-pre-commit-stash

[the remark this answers is not in the document]
Comment (marko-radosavljevic) [reply scripts/pre-commit:19]: https://user-images.githubusercontent.com/30927961/160714683-de9753ef-4829-4fae-8c82-171ae3e683b9.mp4

[the remark this answers is not in the document]
Comment (marko-radosavljevic) [reply scripts/pre-commit:19]: Just me playing around and trying a few basic things. I just pulled, changed to this branch, nothing else. `cas` is alias for `git add -A && git commit -S -m`, changes were the same as in the first example.

**Review by marko-radosavljevic — COMMENTED** [review]
Since we decided to go with a pre-commit hook, we don't need a previous `spotlessApply` on compile/build strategy.

That way, everyone can enjoy their own formatting, without getting annoyed with constant spotless re-formatting while they work.

If you agree, you can remove this line:

https://github.com/Together-Java/TJ-Bot/blob/3ea40dc8e063f3e46455d30d219cb4960626268c/build.gradle#L58

File: scripts/pre-commit (hunk 1 above)
[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit:19]: nice observation

File: scripts/pre-commit [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+echo "****Done pre-commit hooks****"
```
[the remark this answers is not in the document]
Comment (borgrel) [reply scripts/pre-commit:20]: newlines are useful and making reading code better
if u want to spread a command over multiple lines in bash while still keeping it as a single command u prepend a \ before the newline

File: scripts/pre-commit [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#!/bin/sh
+
+echo "****Running pre-commit hooks****"
+# Keep staged changes and put unstaged aside for the hook (bypass hook to avoid recursion)
```
[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit:4]: i dont quite get what u mean with this.

[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit:4]: I mostly had issues understanding your sentence, what it means. But now its clear, thanks.

[the remark this answers is not in the document]
Comment (marko-radosavljevic) [reply scripts/pre-commit:4]: lmao yeah, in short: try to document it better if you can, so it doesn't shoot us in the foot down the line. :D

Shell scripts are few orders of magnitude more dangerous than regular java code, and we run it before each commit.

File: scripts/pre-commit (hunk 1 above)
[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit]: I think I fixed the issues. If possible, please test it again - that would be very helpful, thank you very much!

File: scripts/pre-commit (hunk 3 above)
[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit]: Added extensive docs, thanks

File: scripts/pre-commit [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+# (otherwise spotless would apply on them as well, see https://github.com/diffplug/spotless/issues/623).
+# There are a few edge cases to handle here; for example if there are no unstaged changes, stash would not create a stash.
+# Hence, the following pop of the stash must be guarded (see https://stackoverflow.com/a/20480591/2411243).
+
+# Mini commit for staged changes, so that we can stash away unstaged in the next step (bypass hook with "no-verify" to avoid recursion)
+git commit --no-verify --message "WIP"
+# Put unstaged away
+create_stash
+# Get back staged changes
+git reset --soft HEAD^
```
[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit:29]: For real, why is there no proper copy-paste script somewhere that does exactly what we want... 😆 Thanks for being so careful with reviewing this though.

[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit]: fixed, please check again, thanks

File: scripts/pre-commit [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+# WIP-commit for the staged changes and put the unstaged changes aside using stash
+# (otherwise spotless would apply on them as well, see https://github.com/diffplug/spotless/issues/623).
+# There are a few edge cases to handle here; for example if there are no unstaged changes, stash would not create a stash.
+# Hence, the following pop of the stash must be guarded (see https://stackoverflow.com/a/20480591/2411243).
+
+# Mini commit for staged changes, so that we can stash away unstaged in the next step (bypass hook with "no-verify" to avoid recursion)
+if ! git commit --no-verify --message "WIP"; then
+  # Commit failed (for example if there is nothing staged), early-out
+  exit $?
+fi
```
[the remark this answers is not in the document]
Comment (Zabuzard) [reply scripts/pre-commit]: Good stuff. Fixed, please check again, thanks.

**Review by marko-radosavljevic — APPROVED** [review]
Looks good. Thanks for documenting it nicely. :relaxed: :heart:

Consider this PR approved even if you force push some history cleanup.

**Pull request merged**